### ISSUE TITLE
[2.8.0] - Running out of memory on false(y) expectation.

### ISSUE THREAD

**Issue by elliotwesoff** [jasmine/jasmine#1422]
Hello,

I'm trying to run an expectation against a variable containing a DOM element for my Angular 4 specs, and one of the specs are causing the browser to go into an infinite loop and crash. I stumbled upon this when I changed the implementation of my tests. Where the `element` variable used to be null (see example below), it's now present on the DOM and is an instance of the `DebugElement` class. When it hits the expectation line it freezes, and with Task Manager open I can watch my computer's memory usage skyrocket (I actually let it go a bit too long once and my screen went black and had to reboot).

## Expected Behavior
The test suite should not freeze, and my computer should not run out of memory and crash.

## Current Behavior
The test suite freezes at the current spec and fills up my computer's memory until it crashes.

## Possible Solution
I haven't gone through the jasmine source yet, so I'm not sure :/

## Suite that reproduces the behavior (for bugs)
```
it('should not crash the computer', () => {
    let element = fixture.debugElement.query(By.css('div#container'));
    expect(element).toBeFalsy();
    expect(element).toBe(null); // this makes it crash as well...
});
```

## Your Environment

* Version used: jasmine-core@2.8.0
* Environment name and version (e.g. Chrome 39, node.js 5.4): Chrome 61.0.3163.100, Angular 4.0.0
* Operating System and version (desktop or mobile): Windows 10 Professional
* Link to your project: It's a private repository, sorry.

**Comment by scottmcmaster** [comment]
Also seeing something like this. Right offhand, it looks like maybe a stack overflow in the prettyprinting of the object for the failure message.

**Comment by sgravrock** [comment]
The pretty printer tries to prevent both stack overflows and memory exhaustion, but it's possible that your DOM tree is both deep and wide enough to get past its checks. You might be able to fix the problem by setting `jasmine.MAX_PRETTY_PRINT_DEPTH` to a smaller value than the default of 40. You might find that the matcher failure message isn't very useful, though, since it's likely to contain a serialization of a lot more `DebugElement` instances than you might expect. You might actually find an assertion like `expect(!!element).toBeFalsy()` to be more useful.

**Comment by akaztp** [comment]
Got me one hour to pinpoint to this problem.
What is more interesting is that 
```expect(element).not.toBe(null);```
does not have any problem

**Comment by sgravrock** [comment]
I've made a couple of changes on master that should help with this. I've reduced `jasmine.MAX_PRETTY_PRINT_DEPTH` to 8. Jasmine will also now stop pretty-printing after a certain number of characters, configurable as `jasmine.MAX_PRETTY_PRINT_CHARS`, even if it hasn't hit either `MAX_PRETTY_PRINT_DEPTH` or `MAX_PRETTY_PRINT_ARRAY_LENGTH`.

**Comment by slackersoft** [comment]
With the new default values for the pretty printer and the inclusion of a maximum number of characters to print, this issue should be solved in 2.9 and 3.0. 3.0 also includes better pretty printing of DOM nodes which should help this problem as well.

I'm going to close this. Thanks for using Jasmine!
